src/tools/cardos5-tool.c

```diff
@@ -0,0 +1,1245 @@
+	*mac_len = 8;
+
+	return 0;
+#else
+	warnx("%s: AES-CMAC support missing in OpenSSL", __func__);
+	return -1;
+#endif
+}
+
+/*
+ * Perform AES CBC as described in CardOS 5.0 User's Manual. The relevant
+ * sections and figures are: 2.6.2 (pg. 89), 2.6.2-2 (pg. 91), and 4.1.1 in the
+ * Packages & Release Notes (PRN).
+ *
+ * We receive a system key (which is the StartKey) and use the invariant seed
+ * value to derive an encryption key from it. This derivation process is
+ * underdocumented and therefore unclear; the only hint in the manual being
+ * "SMx4H_ENC_KEY = ENC_SYSTEMKEY(SEED)". It does not state which
+ * initialisation vector (IV) should be used in this calculation.
+ *
+ * Once in possession of the encryption key, figure 2.6.2-2 (referenced above)
+ * suggests that we need to encrypt 16 zero bytes to get the IV to be used when
+ * encrypting the data payload.
+ */
+int
+aes_cbc(const uint8_t *data, size_t data_len, const uint8_t *syskey,
+    size_t syskey_len, uint8_t **out, size_t *out_len)
+{
+	AES_KEY		 k;
+	uint8_t		 zero_iv[AES_BLOCK_SIZE];
+	uint8_t		 unencrypted_iv[AES_BLOCK_SIZE];
+	uint8_t		 encrypted_iv[AES_BLOCK_SIZE];
+	uint8_t		 enckey[AES_BLOCK_SIZE];
+	uint8_t		*seed;
+	ssize_t		 seed_len;
+
+	/*
+	 * Step 1: Derive encryption key from system key using the invariant
+	 * seed.
+	 */
+
+	load_file(seed_path, 16, 16, &seed, &seed_len);
+	memset(unencrypted_iv, 0, sizeof(unencrypted_iv));
+	memset(zero_iv, 0, sizeof(zero_iv));
+
+	AES_set_encrypt_key(syskey, syskey_len * NBBY, &k);
+	AES_cbc_encrypt(unencrypted_iv, encrypted_iv, sizeof(encrypted_iv), &k,
+	    zero_iv, AES_ENCRYPT);
+	AES_cbc_encrypt(seed, enckey, sizeof(enckey), &k, encrypted_iv,
+	    AES_ENCRYPT);
+
+	/*
+	 * Step 2: Use encryption key to encrypt the IV.
+	 */
+
+	memset(unencrypted_iv, 0, sizeof(unencrypted_iv));
+	memset(zero_iv, 0, sizeof(zero_iv));
+
+	AES_set_encrypt_key(enckey, sizeof(enckey) * NBBY, &k);
+	AES_cbc_encrypt(unencrypted_iv, encrypted_iv, sizeof(encrypted_iv), &k,
+	    zero_iv, AES_ENCRYPT);
+
+	/*
+	 * Step 3: Encrypt the payload.
+	 */
+
+	*out = calloc(data_len, sizeof(char));
+	if (*out == NULL)
+		return -1;
+
+	*out_len = data_len;
+	AES_cbc_encrypt(data, *out, data_len, &k, encrypted_iv, AES_ENCRYPT);
+
+	return 0;
+}
+
+/* ISO padding for blocksize m adds 1 to m bytes */
+#define ISO_PAD_LEN(u, m)	((m) - ((u) % (m)))
+#define BLOCK_SIZE		8
+
+int
+add_iso_pad(uint8_t *buf, size_t buflen, size_t datalen, size_t m)
+{
+	size_t i;
+
+	if (buflen % BLOCK_SIZE != 0)
+		return -1;
+
+	/* We always add at least one byte */
+	if (buflen < datalen + 1)
+		return -1;
+	if (datalen + ISO_PAD_LEN(datalen, m) != buflen)
+		return -1;
+
+	/* First pad byte is 0x80, rest is 0x0 */
+	buf[datalen] = 0x80;
+	for (i = datalen + 1; i < buflen; i++)
+		buf[i] = 0x0;	/* obsolete if buffer was calloc'd */
+
+	return 0;
+}
+
+#define MAC_HDR		4
+
+void
+smmodex4h(const uint8_t *data, size_t data_len, uint8_t *key, size_t key_len)
+{
```

Comment:
have you thought about directly implementing the `sm_ops` and using `card->sm_ctx.sm_mode = SM_MODE_TRANSMIT`. This would solve SM at the card driver level which would maybe also allow other uses of it.